### PR TITLE
Update __exportStar helper to skip default and __esModule members

### LINKED ISSUE #37234: Default export conflicts with star (regression)
**TypeScript Version:**  3.7.x-dev.201xxxxx

**Search Terms:**

**Code**

```ts
// src/export.ts
export { GUI as Pjax, GUI as default } from './layer/interface/service/gui';
export { router } from './lib/router';
// index.ts
export * from './src/export';
export { default } from './src/export';
```

```js
Object.defineProperty(exports, "__esModule", { value: true });
// Define the default export.
__exportStar(_dereq_("./src/export"), exports);
var export_1 = _dereq_("./src/export");
// Redefine the default export and then this operation causes an exception.
Object.defineProperty(exports, "default", { enumerable: true, get: function () { return export_1.default; } });
```
You can reproduce the error by reverting https://github.com/falsandtru/pjax-api/commit/f814698a688feb449f49a086259538d17066f30e#diff-ed009b6b86b017532ef0489c77de5100L2-R2
**Expected behavior:**
pass
**Actual behavior:**
Uncaught TypeError: Cannot redefine property: default
**Playground Link:** 

**Related Issues:**

### PR DESCRIPTION
**Pull request by weswigham**
Fixes #37234

### CODE REVIEW

File: src/compiler/transformers/module/module.ts [hunk 1]
```diff
@@ -1886,7 +1886,7 @@ var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (
         priority: 2,
         text: `
             var __exportStar = (this && this.__exportStar) || function(m, exports) {
-                for (var p in m) if (!exports.hasOwnProperty(p)) __createBinding(exports, m, p);
+                for (var p in m) if (p !== "default" && p !== "__esModule" && !exports.hasOwnProperty(p)) __createBinding(exports, m, p);
```
Comment (rbuckton) [inline src/compiler/transformers/module/module.ts]: Won't `__esModule` always be skipped because we emit it unconditionally anyways?

**Comment by bluelovers** [comment]
i think this not a good way for avoid bug

u will see bug happen again when run e1.ts

e1.ts
```ts
export * from './e2'

export function abc()
{

}

```

e2.ts
```ts
export function abc()
{

}
```

**Comment by weswigham** [comment]
@bluelovers We [hoist an undefined member assignment](http://www.typescriptlang.org/v2/en/play?module=1&ts=next#code/KYDwDg9gTgLgBAKjgMyhAtnA5AOgPTABMWAUCaJLCgK4B2AxjAJYS1wCGARvQBQCUJAN5kAviThwgA) so that actually works just fine (as the hoisted assignment blocks us from overwriting it with the `exportStar` helper). It's only `default` that's a little different, since we don't track `default` as an exported name. That may actually be a better fix - hoisting an `exports.default = void 0` assignment would be a bit more inline with other exports, at least.

**Comment by bluelovers** [comment]
oh, i see

it already fix in current next version

my version still at `20200228`

**Comment by rbuckton** [comment]
@weswigham no, I don't think we should hoist an assignment for `default`, as that is observable on the module object in CommonJS (and wouldn't be observable on the module namespace object in a true ES module).

I think the behavior of the helper to explicitly skip `default` is correct and consistent with the spec behavior: https://tc39.es/ecma262/#sec-getexportednames (Step 9.c.i.)

**Comment by bluelovers** [comment]
how about make a `exports.default = void 0;`
when has `export default`

**Comment by weswigham** [comment]
@rbuckton  ?

**Pull request merged**